src/service.rs

```diff
@@ -424,14 +836,410 @@ impl Service {
+                    // When the ticket time expires a new REGTOPIC request is automatically sent to the
+                    // ticket issuer and the registration attempt stays in the [`RegistrationState::Ticket`]
+                    // from sending the first REGTOPIC request to this contact for this topic.
+                    self.reg_topic_request(active_ticket.contact(), active_topic.topic().clone(), RequestTicket::RemotelyIssued(active_ticket.ticket()));
+                }
+                Some(topic_query_progress) = self.active_topic_queries.next() => {
+                    match topic_query_progress {
+                        TopicQueryState::Finished(topic_hash) | TopicQueryState::TimedOut(topic_hash) => {
+                            if let Some(query) = self.active_topic_queries.queries.remove(&topic_hash) {
+                                if let Some(callback) = query.callback {
+                                    if callback.send(Ok(query.results.into_values().collect::<Vec<_>>())).is_err() {
+                                        warn!("Callback dropped for topic query {}. Results dropped", topic_hash);
+                                    }
+                                }
+                            }
+                        }
+                        TopicQueryState::Unsatisfied(topic_hash) => self.send_topic_queries(topic_hash, None),
+                    }
+                }
+                _ = registration_interval.tick() => {
+                    trace!("New registration interval, {}/{} topics to publish", topics_to_reg_iter.clone().count(), self.registration_attempts.len());
+                    let mut sent_regtopics = 0;
+                    let mut topic_item = topics_to_reg_iter.next();
+                    while let Some((topic, _topic_hash)) = topic_item {
+                        trace!("Publishing topic {} with hash {}", topic, topic.hash());
+                        topic_item = topics_to_reg_iter.next();
+                        // It could be that a topic has been set to stop registration since the
+                        // iteration through topics_to_reg_iter was started, in that case skip
+                        // that topic.
+                        if !self.registration_attempts.contains_key(&topic) {
+                            continue;
+                        }
+                        sent_regtopics += self.send_register_topics(topic.clone());
+                        if sent_regtopics >= MAX_REGTOPICS_REGISTER_PER_INTERVAL {
+                            break
+                        }
+                    }
+                    if topics_to_reg_iter.next().is_none() {
+                        topics_to_reg_iter = self.registration_attempts.keys().map(|topic| (topic.clone(), topic.hash())).collect::<Vec<(Topic, TopicHash)>>().into_iter();
+                    }
+                }
+            }
+        }
+    }
+
+    fn get_active_topics(&mut self) -> HashMap<Topic, Vec<NodeId>> {
```

Comment:
do you think this kind of maintenance could be done on an interval for example? in order to remove the mutability of a GET service function. We could still do this maintenance step when the call is done but at least it wouldn't be the only moment in which is done

Reply:
We should also move the maintenance code to it's own function (it's most of the code, yes) but it's repeated at least once more in the `send_register_topics` logic